File: Progammers_다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;
 
int solution(int bridge_length, int weight, vector<int> truck_weights) 
{
    queue<int> bridge;
    int time = 0;
    int sum_weight = 0;
    int truckIdx = 0;
    
    while(1)
    {    
        time++;
        
        // if the queue size == length of bridge, it means trucks arrive
        if(bridge.size() == bridge_length)
        {
            sum_weight -= bridge.front();
            bridge.pop();
        }
        
        // if sum_weight + truckweights is okay
        if(sum_weight + truck_weights[truckIdx] <= weight)
        {
            // if last truck
            if(truckIdx == truck_weights.size() - 1)
            {
                // add time for the last truck
                time += bridge_length;
                break;
            }
            
            // let go truck
            bridge.push(truck_weights[truckIdx]);
            sum_weight += truck_weights[truckIdx];
            truckIdx++;
        }
        // if sum_weight + truckweights is NOT okay, let go 0
        else
        {
            bridge.push(0);
        }    
    }
    
    return time;
}

/* revised from https://rile1036.tistory.com/12 */

File: Progammers_다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    time = 0
    bridge = [0] * bridge_length
    
    while len(bridge) != 0:
        time += 1
        bridge.pop(0)   # let go the very first element 
        
        if truck_weights:   # if truck_weights value exist 
            # if current bridge weight + new truck weight is okay
            if sum(bridge) + truck_weights[0] <= weight:
                # pop truck and push to the bridge
                bridge.append(truck_weights.pop(0))
            else:
                # if it's not ok, push 0
                bridge.append(0)
        
    return time

# revised from https://velog.io/@devjuun_s/%EB%8B%A4%EB%A6%AC%EB%A5%BC-%EC%A7%80%EB%82%98%EB%8A%94-%ED%8A%B8%EB%9F%AD-%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%A8%B8%EC%8A%A4 */